### ISSUE TITLE
Issues with handlers.yaml regarding huss and rlut variables

### ISSUE THREAD

**Issue by TonyB9000** [E3SM-Project/e3sm_to_cmip#216]
I am looking at some (CMIP6) variables, whose names exist at different frequencies (huss, rlut, and pr) in the “handlers.yaml” file
 
First, “huss”:
 
-   name: huss
    units: '1'
    raw_variables: [QREFHT]
    table: CMIP6_Amon.json
    unit_conversion: null
    formula: null
    positive: null
    levels: null
 
There is a table “CMIP6_day.json” that otherwise had the very same info (raw var = QREFHT).  I may be encountering a problem, in that my processing of the CMIP6 output datasets
 
            Amon.huss       (obtained from native “atmos ... mon” data)
            day.huss          (obtained from native “atmos ... day” data)
 
expects the “handler table name” to match the native data frequency (Amon = atmos mon, day = atmos day) and there is no longer a table “CMIP6_day.json” listed to match the day generation.
 
From the “huss” section of CMIP6_Amon.json, and CMIP6_day.json (from https://github.com/PCMDI/cmip6-cmor-tables)
 
            [CMIP6_Amon.json]
        "huss": {
            "frequency": "mon",
            "modeling_realm": "atmos",
            "standard_name": "specific_humidity",
            "units": "1",
            "cell_methods": "area: time: mean",
            "cell_measures": "area: areacella",
            "long_name": "Near-Surface Specific Humidity",
            "comment": "Near-surface (usually, 2 meter) specific humidity.",
            "dimensions": "longitude latitude time height2m",
            "out_name": "huss",
            "type": "real",
            "positive": "",
            "valid_min": "",
            "valid_max": "",
            "ok_min_mean_abs": "",
            "ok_max_mean_abs": ""
        },
 
            [CMIP6_day.json]
        "huss": {
            "frequency": "day",
            "modeling_realm": "atmos",
            "standard_name": "specific_humidity",
            "units": "1",
            "cell_methods": "area: time: mean",
            "cell_measures": "area: areacella",
            "long_name": "Near-Surface Specific Humidity",
            "comment": "Near-surface (usually, 2 meter) specific humidity.",
            "dimensions": "longitude latitude time height2m",
            "out_name": "huss",
            "type": "real",
            "positive": "",
            "valid_min": "",
            "valid_max": "",
            "ok_min_mean_abs": "",
            "ok_max_mean_abs": ""
        },
 
Second: “rlut”:
 
-   name: rlut
    units: W m-2
    raw_variables: [FLUT]
    table: CMIP6_day.json
    unit_conversion: null
    formula: FSNTOA - FSNT + FLNT     <- cannot generate this from “day” native data – no FSNTOA present.  Should this be FLUT (modulo units)?
    positive: up
    levels: null
 
-   name: rlut
    units: W m-2
    raw_variables: [FSNTOA, FSNT, FLNT]
    table: CMIP6_Amon.json
    unit_conversion: null
    formula: FSNTOA - FSNT + FLNT
    positive: up
    levels: null
 
Note that both atmos monthly and atmos day data have a native var “FLUT”.  I had run into a problem where datasm sought “info” for “rlut” (with monthly data, aiming to produce Amon output), found a handler for rlut (CMIP6_day) whose raw variable “FLUT” was indeed in the monthly data, and used that handler instead of the (assumed correct) Amon handler formula.  So, it passed “FLUT” as the variable, along with FSNTOA, FSNT, FLNT to apply, instead of only FSNTOA, FSNT, FLNT.  The regridder happily accepted the FLUT task, and the happily accepted that output.

Trying to fix this, I added the following to “run_info_mode (3)”:
 
                    if self.freq == "mon" and handler['table'] == "CMIP6_day.json":
                        continue
                    if self.freq == "day" and handler['table'] == "CMIP6_Amon.json":
                        continue
 
But this causes “day.huss” to fail, because there is no “CMIP6_day.json” for self.freq == "day" listed in the handlers.yaml file.

**Comment by beharrop** [comment]
I have recently encountered this issue again while processing some data using e3sm_to_cmip.  If `FLUT` is available in the monthly time series, then the corresponding `rlut` output will take all its metadata from the `CMIP6_day.json` file.  There aren't any errors raised by this process, and the data inside rlut_day_<more_stuff>.nc file are still the monthly data, just with the wrong file name and metadata.

**Comment by TonyB9000** [comment]
@beharrop Thanks for reporting this!  Clearly, this needs to be addressed, and possible remediation made.

**Comment by TonyB9000** [comment]
@beharrop  @chengzhuzhang As of late last night (20250402), I believe we have completed the publication of all monthly (154), daily (60) and hourly (8) datasets, both to the ALCF (Argonne) and LLNL index servers.

If any thing amiss arises or is an outstanding issue, please keep it alive by pulsing us.  We try to juggle as many balls as possible.

I believe we need to revisit the E2C handling for variable that appear in multiple tables (have multiple handlers).  I defer to Jill on priorities here.

**Comment by TonyB9000** [comment]
@chengzhuzhang Indeed.  I will want to itemize the full range of (tables)X(variables) for v3 into machine-parsable lists, which are 2D or 3D, and in particular identify any whose name appears in multiple tables.  I believe that would be the focus of work on this issue.

**Comment by beharrop** [comment]
Thanks @TonyB9000, @chengzhuzhang.  A quick glance at LLNL and ALCF looks like all is well.

**Comment by tomvothecoder** [comment]
Hey @TonyB9000 can you provide the minimum `e3sm_to_cmip` command/script to reproduce this issue?

**Comment by tomvothecoder** [comment]
I will try the recommendation below in a new pull request.

### Recommendation (according to ChatGPT):

To fully resolve this and make the tool robust:

1. **Add a frequency check when selecting handlers**:

   * Cross-reference `self.freq` with the expected frequency in the CMIP6 table metadata (e.g., `"frequency": "mon"`).

2. **Optionally: Add a frequency field in `handlers.yaml`**:

   * This avoids relying solely on the table name and enables more precise handler filtering.

3. **Log a warning or raise an error** when multiple handlers match the same variable name but apply to different frequencies or formulas.

**Comment by tomvothecoder** [comment]
Hi @TonyB9000, just following up to my above [comment](https://github.com/E3SM-Project/e3sm_to_cmip/issues/216#issuecomment-2880852926) for your help. Thanks!

**Comment by TonyB9000** [comment]
@tomvothecoder Sorry - I haven't been keeping up on this issue.  Some 2-3 years ago, I noted that certain variable names appeared in different tables (on frequencies) and attempted to revamp the handler selection process (this involved returning a list of handlers or tables for the e2c "--info" mode operation).  I think ChatGPT "understands" the issue ;).    My partial solution was to have a "list" of outputs written (even though 99% of the time, the list would only have 1 entry) and have the calling process decide which to use.  Something more elegant (more direct, less wasteful) would be best.

**Comment by tomvothecoder** [comment]
> [@tomvothecoder](https://github.com/tomvothecoder) Sorry - I haven't been keeping up on this issue. Some 2-3 years ago, I noted that certain variable names appeared in different tables (on frequencies) and attempted to revamp the handler selection process (this involved returning a list of handlers or tables for the e2c "--info" mode operation). I think ChatGPT "understands" the issue ;). My partial solution was to have a "list" of outputs written (even though 99% of the time, the list would only have 1 entry) and have the calling process decide which to use. Something more elegant (more direct, less wasteful) would be best.

@TonyB9000 No problem. Can you provide a minimum script/command to reproduce this behavior?


> Instead of cycling over all handlers seeking a match, perhaps a table could be maintained so that a direct lookup could be applied. (Maintaining the look-up table would be a chore, pushed to git whenever handlers change.) Just a thought.

If I understand your suggestion correctly, we actually have a "look-up" table already ([handlers.yaml](https://github.com/E3SM-Project/e3sm_to_cmip/blob/master/e3sm_to_cmip/cmor_handlers/handlers.yaml)), which stores variable handler definitions (besides MPAS variables).

The current process involves creating a list of handlers by converting each yaml entry as a `VarHandler` object, then parsing that list of handlers for matching ones. As far as **performance**, I did not measure speed it takes to parse the list of handlers for matches. This would be a good experiment to tackle whenever you or I have time.

**Comment by TonyB9000** [comment]
@tomvothecoder Is the table (handlers.yaml) complete?  I thought there might be some exceptions.

If complete, that would be the way to go, assuming "same-var, different frequency" are distinguished.

(ASIDE:  I am guessing that the "converting each yaml entry as a VarHandler object," occurs whenever someone issues "e3sm_to_cmip --help" or "e3sm_to_cmip --version", and is why it takes like a minute to return an answer.  I think it is part of the "run" class init, and occurs before arg-parsing is completed and reviewed for short-cuts.
